### PR TITLE
[collective] Add a torch.compile + functional_collectives test

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109647
* __->__ #110688

Add a test to ensure functional_collectives + torch.compile always works.

Differential Revision: [D50001491](https://our.internmc.facebook.com/intern/diff/D50001491/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/distributed/test_functional_api.py [hunk 1]
```diff
@@ -458,6 +458,16 @@ def test_all_gather_into_tensor_coalesced(self):
         self.assertEqual(torch.ones([4 * dist.get_world_size()]), res[0])
         self.assertEqual(torch.ones([4 * dist.get_world_size()]) + 1, res[1])
 
+    @skip_if_lt_x_gpu(WORLD_SIZE)
+    @requires_nccl()
+    @with_comms()
+    def test_tracing(self):
```
Comment (wanchaol) [inline test/distributed/test_functional_api.py:464]: I am a bit surprised this issue was not captured in CI, we have bunch of tests should guard it https://github.com/pytorch/pytorch/blob/main/test/distributed/test_inductor_collectives.py#L572

but thanks for adding more tests to guard it!

**Review by wanchaol — APPROVED** [review]
Let's try to land this to temporarily unblock, and @yanboliang can address this fundamentally in a separate PR

File: test/distributed/test_functional_api.py [hunk 2]
```diff
@@ -544,6 +564,17 @@ def test_all_to_all_single_split_sizes_none(self):
         expected = torch.cat(expected)
         self.assertEqual(y, expected)
 
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @skip_if_lt_x_gpu(WORLD_SIZE)
```
Comment (fduwjj) [inline test/distributed/test_functional_api.py:568]: Well this may until run for the periodic label enabled.

**Review by fduwjj — APPROVED** [review]
LGTM

File: test/distributed/test_functional_api.py (hunk 2 above)
Comment (fegin) [reply test/distributed/test_functional_api.py:568]: WORLD_SIZE is 2, so I guess it is fine?

Comment (fduwjj) [reply test/distributed/test_functional_api.py:568]: Oh, NVM, I thought it was 4.
